### ISSUE TITLE
Improve Custom Resources handling

### ISSUE THREAD

**Issue by metacosm** [fabric8io/kubernetes-client#2611]
While some improvements have been made to how Custom Resources (CRs) are handled, some issues still remain. 

-  #2609 fixes an issue where a specific list implementation was required to be able to use `customResources(Class customResourceClass)`. However, an issue still exist where `CustomResourceDefinitionContext.fromCustomResourceType` [cannot properly infer the needed group and version](https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/CustomResourceDefinitionContext.java#L64-L65) unless the the implementation explicitly provide them. 
- There is still boilerplate around CRs and some of this could be alleviated by adding more specific support similar to what has been done in the [Istio Java API project](https://github.com/snowdrop/istio-java-api) and is currently explored in [Dekorate](https://github.com/dekorateio/dekorate/pull/674). It would make sense to unify these efforts and bring them in the client directly so that they could benefit downstream projects more widely.

Regarding the first issue, in the short term, we could require custom resource classes to extend `CustomResource` and remove the default implementation that this class provide of the `getApiVersion` method so that implementations would need to provide a value, thus ensuring that `fromCustomResourceType` would properly be able to identify the CR on the cluster.

**Comment by manusa** [comment]
I didn't have a chance to check what's done either on IstioJC or Dekorate. I'm guessing for the latter it's something annotation based.

From my yet naive perspective, I see several options (not sure if they are even possible or make sense altogether).

- Static code generation tool that generates a "client" for a given set of CRD yamls.
- Or runtime analysis of custom annotated POJOs to generate the corresponding API calls in a multi-purposed client. I'm not sure if this would be problematic with Quarkus or GraalVM, but it's certainly the way to go for me.
  We already have some work done on this area.

**Comment by metacosm** [comment]
@manusa could you detail the second option, please? Wouldn't you generate code at build time as opposed to runtime?

The Istio Java API approach uses the following approach: a generative pipeline is set up in a similar way to what the client is doing (i.e. go -> json schema -> jsonschema2pojo annotator -> sundrio). The difference is that for go structs that are marked as k8s spec structs, we annotate them with specific annotations and generate related classes (CR class itself, Status and List) automatically based on velocity templates.

Dekorate uses similar ideas except that we already have the java code, so we just need to annotate them and let the generator do the work. Marking a Spec class as a CR class would then trigger automated generation on CR, Status and List classes associated with the Spec class, which is, most of the time, the only thing that is not boilerplate. From there, we could also generate the CRD yaml and even a custom client to deal with the CR.

My question, though, is more related to how you would conceive the workflow from a Java developer's perspective? Would simply creating the Spec class and annotating it work?

**Comment by manusa** [comment]
So to clear things up, I see 2 possible scenarios that have nothing to do with each other.

The first scenario is when you want to consume a **third party CRD based API** (e.g. Istio). Then, my first assumption applies, you want to go from the CRD yaml files to a Java client (GoStructs?>YAML>Java). Since you expect to have types for your resources, this can only work as a static generator (a K8s `wsimport` of sorts ;)) (See https://github.com/rohanKanojia/fabric8-gostruct2pojogenerator and related discussion).


The second scenario would be a more **Java-first approach**. e.g. Given you want to build an operator for your Java application, maybe you want to define your own CRDs for your application in plain Java. So you want to go the other way around (POJO>YAML/k8s).
Another scenario for this strategy could be that you want to partially consume a third-party API without having to deal or worry about all of their types.

We currently support something like this with [`RawCustomResourceOperationsImpl`](https://github.com/fabric8io/kubernetes-client/blob/7beda8dd2449284393318bdf7a4c7e9adec6b942/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java#L237) (typeless) and [`CustomResourceOperationsImpl`](https://github.com/fabric8io/kubernetes-client/blob/c7acc9feebe97f46831ed4c7153e479ead0a28ce/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/DefaultKubernetesClient.java#L424) (typed). 
Both you and Ioannis already provided enhancements that improve the experience in this direction (#2606 and #2609).
What I feel we're missing in order to close the circle is the CRD generation and "deployment", so that everything is Java based.

_"As a Java developer I want to create a Kubernetes Operator only by using Java. I will provide a set of Java types that represent my model and Fabric8 will infer everything needed to deploy my CRD and interact with the extended API."_

**Comment by metacosm** [comment]
@manusa thanks! 👍

**Comment by rohanKanojia** [comment]
I'm just sharing some feedback received from @FWiesner a few months ago(when discussing https://github.com/fabric8io/kubernetes-client/issues/2489) . He suggested moving CustomResourceDefinitionSpec related configuration to annotations like this:
```java
package io.fabric8.kubernetes.examples.crds;

import io.fabric8.kubernetes.api.model.Namespaced;
import io.fabric8.kubernetes.client.CustomResource;
import io.fabric8.kubernetes.model.annotation.ApiGroup;
import io.fabric8.kubernetes.model.annotation.ApiVersion;
import io.fabric8.kubernetes.model.annotation.Plural; // Doesn't exist currently in client

@ApiGroup("io.fabric8.examples")
@ApiVersion("v1alpha1")
@Plural("foos")
public class Foo extends CustomResource implements Namespaced {
  // ...
}
```

These annotations then could be read by this method(defaulting to the values we're having right now in the method)
https://github.com/fabric8io/kubernetes-client/blob/12e17797d93ce4c85115174032d39186bd9bc67c/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/CustomResourceDefinitionContext.java#L56

What do you think about this?

**Comment by iocanel** [comment]
This makes sense to me. We are already using these exact annotations in downstream projects like the `istio-java-api` and we could possibly reuse them in dekorate where we generate crd from pojos.

**Comment by metacosm** [comment]
Should we take the opportunity to remove some boiler plate code as well from the generated classes? If we use default implementations on `HasMetadata`, we could remove some of the properties (kind, apiVersion) and their associated accessors.

**Comment by metacosm** [comment]
I've started working on some improvements in #2661. More to come.

**Comment by metacosm** [comment]
Ideally, I'd like us to provide a `CustomResourceDefinition.buildFrom(CustomResource)` method but I'm not sure what's the best way to do it. The easiest way would be to introduce a `CustomResourceDefinitionUtils` class that would provide such a method but I'm afraid it wouldn't really be easily discoverable. What do you think, @manusa @iocanel @rohanKanojia ?

**Comment by metacosm** [comment]
I guess we could put it on `CustomResourceDefinitionContext` but that feels kind of weird… 😄

**Comment by manusa** [comment]
> Ideally, I'd like us to provide a CustomResourceDefinition.buildFrom(CustomResource) 

This really makes sense from the operator perspective.
Defining the complete CR(D) API via annotated POJOs seems to me the way to accomplish this too (Having to define POJO + CRD seems an unnecessary overhead for a **new** API).
We'll probably need a few more annotations to be able to define a complete spec or add compatibility with some other JSR (e.g. JSR 303, etc.)

Regarding the location, IMO this doesn't belong to the `CustomResourceDefinition` class either (since there are already 2 of them). Maybe we should start looking for a dedicated package where all of these tools and utils are located.
From an Operator framework perspective, I see that eventually you will not want to call these methods via the published API, but to automatically scan packages and generate whatever is necessary in a more transparent configuration-less way. So if you have some annotated POJOs in your project, things should be readily available and provided for you (I'm thinking of something analogous to a JPA `@Entity`)

**Comment by metacosm** [comment]
One thing that I'm not too sure about either is how to deal with multiple versions…

**Comment by manusa** [comment]
> One thing that I'm not too sure about either is how to deal with multiple versions…

The aforementioned "toolset" should include an utility method that decided this for you depending on the target cluster and API availability. That's another reason why I think CRD generation should be less imperative (i.e. tool checks for CRD deployment in cluster, tool deploys CRD if necessary using latest CRD API available, ...)

**Comment by metacosm** [comment]
Based on the PR, what else (if anything) do you think we should be doing to improve CR handling, @manusa @iocanel @rohanKanojia ?

**Comment by metacosm** [comment]
> > One thing that I'm not too sure about either is how to deal with multiple versions…
> 
> The aforementioned "toolset" should include an utility method that decided this for you depending on the target cluster and API availability. That's another reason why I think CRD generation should be less imperative (i.e. tool checks for CRD deployment in cluster, tool deploys CRD if necessary using latest CRD API available, ...)

Could you elaborate on this, please?

**Comment by manusa** [comment]
> Could you elaborate on this, please?

So far we've worked on the CRD>CR consumption part. We've provided the tools to generate a CRD API client that works with any CRD API version (v1beta1 or v1). The main reason for this is that CRDs comply to a standard RESTful API with strong conventions around resource naming and location. Instead of working with CRD instances to declare this client we use a more abstract `CustomResourceDefinitionContext` which contains the required fields to be able to interact with via REST with Kubernetes. (So we could say that `CustomResourceOperations` implementation is more or less complete. :heavy_check_mark: 

However, there's still another requirement/feature (and maybe this belongs more to the SDK) "As a Kubernetes Java developer I want to infer the Custom Resource Definitions for my Java objects automatically".
Considering an abstract `CusomtResourceDefinitionOperations` interface (has the standard CRUD methods) which would have implementations for every API version (v1beta1 or v1), we could define a client method `<T extends CusomtResourceDefinitionOperations> T crdClientFrom(Class<? extends CustomResource> customResourceClazz)` (A).
The method implementation logic should run as follows:
- Check latest CRD API version available in the connected cluster
- Return applicable `CusomtResourceDefinitionOperations` interface implementation instance

Besides, another method to infer a CRD instance from a POJO would be good to have (again, this may probably fit better within the SDK project). This is trickier (probably needs a few more iterations), because there are (as of now) 2 types that represent CustomResourceDefinition. It would make sense some method such as `<T extends HasMetadata> T crdFrom(Class<? extends CustomResource> customResourceClazz)` (B), depending on the APIs available on the cluster, a v1 or v1beta1 CRD instance is returned.
The implementation for this method is were we should be really creative in terms of inferring OpenAPIV3Schema from the POJO fields, annotations, JSR 303, etc.

With both A & B in place, something like this should be feasible `kubernetesClient.crdClientFrom(MyCustomResource.class).create(CRDToolset.crdFrom(MyCustomResource.class));` (or any other CRUD operation on the represented CRD k8s resource).


_(IMHO)_
